File: internal/escalate/escalate.go
package escalate

import (
	"fmt"
	"os/exec"
	"strings"

	"github.com/Pernek-Enterprises/dispatch/internal/config"
	"github.com/Pernek-Enterprises/dispatch/internal/log"
)

// Notify sends an escalation message to the human via OpenClaw.
// Uses `openclaw agent --deliver --channel <ch> --reply-to <target>`.
func Notify(cfg *config.Config, jobID, taskID, message string) error {
	channel := cfg.Notifications.Escalation
	target := cfg.Notifications.Target
	if target == "" {
		target = cfg.Notifications.Channel // backward compat
	}
	if channel == "" {
		log.Warn("No escalation channel configured — logging only")
		log.Info("ESCALATION [%s]: %s", jobID, message)
		return nil
	}

	body := fmt.Sprintf("🚨 **Dispatch Escalation**\n\n**Task:** %s\n**Job:** %s\n\n%s", taskID, jobID, message)

	args := []string{
		"agent",
		"--message", body,
		"--deliver",
		"--channel", channel,
	}
	if target != "" {
		args = append(args, "--to", target)
	}

	binary := findOpenClaw()
	log.Info("Escalating via %s → %s (target: %s)", binary, channel, target)

	cmd := exec.Command(binary, args...)
	out, err := cmd.CombinedOutput()
	if err != nil {
		log.Error("Escalation failed: %v\n%s", err, strings.TrimSpace(string(out)))
		return fmt.Errorf("escalation failed: %w", err)
	}

	log.Info("Escalation delivered for job %s", jobID)
	return nil
}

// NotifyReady sends a notification when a task reaches the human review step.
func NotifyReady(cfg *config.Config, jobID, taskID string) error {
	msg := fmt.Sprintf("✅ **Task ready for review**\n\nThe pipeline has completed and is waiting for your approval.\n\nRun `dispatch answer --job %s \"approved\"` to close it out.", jobID)
	return Notify(cfg, jobID, taskID, msg)
}

// NotifyFailure sends a failure notification.
func NotifyFailure(cfg *config.Config, jobID, taskID, reason string) error {
	msg := fmt.Sprintf("❌ **Job failed**\n\n%s", reason)
	return Notify(cfg, jobID, taskID, msg)
}

// NotifyMaxIterations sends a max-iterations-reached notification.
func NotifyMaxIterations(cfg *config.Config, taskID, step string, iterations int) error {
	msg := fmt.Sprintf("🔁 **Review loop exhausted** — step `%s` hit max iterations (%d).\nNeeds human decision to continue or close.", step, iterations)
	return Notify(cfg, "", taskID, msg)
}

func findOpenClaw() string {
	if p, err := exec.LookPath("openclaw"); err == nil {
		return p
	}
	candidates := []string{
		"/opt/openclaw/bin/openclaw",
		"/usr/local/bin/openclaw",
	}
	for _, c := range candidates {
		if _, err := exec.LookPath(c); err == nil {
			return c
		}
	}
	return "openclaw"
}

File: escalate.ts
import * as child_process from "child_process";
import { type Config } from "./config.js";
import { log } from "./log.js";

function findOpenClaw(): string {
  try {
    return child_process.execFileSync("which", ["openclaw"], { encoding: "utf8" }).trim();
  } catch { /* not in PATH */ }
  for (const c of ["/opt/openclaw/bin/openclaw", "/usr/local/bin/openclaw"]) {
    try { child_process.execFileSync("test", ["-f", c]); return c; } catch { /* skip */ }
  }
  return "openclaw";
}

function notify(cfg: Config, jobId: string, taskId: string, body: string): void {
  const target = cfg.notifications.target || cfg.notifications.channel || "";
  const channel = cfg.notifications.escalation;

  if (!target && !channel) {
    log.warn(`No escalation target configured — logging only: [${jobId}] ${body}`);
    return;
  }

  const args = ["message", "send", "--target", target, "--message", body];
  if (channel) args.push("--channel", channel);

  try {
    child_process.execFileSync(findOpenClaw(), args, { encoding: "utf8" });
    log.info(`Escalation delivered for job ${jobId}`);
  } catch (e) {
    log.error(`Escalation failed for ${jobId}: ${e}`);
  }
}

export function notifyReady(cfg: Config, jobId: string, taskId: string): void {
  const msg = [
    `✅ **Task ready for review**`,
    ``,
    `The pipeline has completed and is waiting for your approval.`,
    ``,
    `Run: \`dispatch answer --job ${jobId} "approved"\` to close it out.`,
  ].join("\n");
  notify(cfg, jobId, taskId, `🚨 **Dispatch** — **Task:** ${taskId} | **Job:** ${jobId}\n\n${msg}`);
}

export function notifyFailure(cfg: Config, jobId: string, taskId: string, reason: string): void {
  const body = `🚨 **Dispatch** — **Task:** ${taskId} | **Job:** ${jobId}\n\n❌ **Job failed**\n\n${reason}`;
  notify(cfg, jobId, taskId, body);
}

export function notifyDeliverableRetry(cfg: Config, jobId: string, taskId: string, step: string, missing: string[], attempt: number, maxAttempts: number): void {
  const body = [
    `⚠️ **Dispatch** — **Task:** ${taskId} | **Step:** ${step}`,
    ``,
    `🔄 **Deliverable retry ${attempt}/${maxAttempts}** — agent didn't produce all required files.`,
    ``,
    `Missing: ${missing.map(f => `\`${f}\``).join(", ")}`,
    ``,
    `Re-dispatching automatically...`,
  ].join("\n");
  notify(cfg, jobId, taskId, body);
}

export function notifyMaxIterations(cfg: Config, taskId: string, step: string, maxIter: number): void {
  const body = `🚨 **Dispatch** — **Task:** ${taskId}\n\n🔁 **Review loop exhausted** — step \`${step}\` hit max iterations (${maxIter}).\nNeeds human decision to continue or close.`;
  notify(cfg, "", taskId, body);
}
